### PR TITLE
[Enhance]  refactor  iou_neg_piecewise_sampler.py

### PR DESCRIPTION
**Pull request by xiliu8006**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily get feedback. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation
Modify iou_neg_piecewise_sampler.py  to support PointRCNN.

## Modification
This modification will do not drop the PartA2  performance，after modifying，the mAP is （79.16 -> 79.27).

### CODE REVIEW

File: mmdet3d/core/bbox/samplers/iou_neg_piecewise_sampler.py [hunk 1]
```diff
@@ -58,8 +58,8 @@ def _sample_neg(self, assign_result, num_expected, **kwargs):
         neg_inds = torch.nonzero(assign_result.gt_inds == 0, as_tuple=False)
         if neg_inds.numel() != 0:
             neg_inds = neg_inds.squeeze(1)
-        if len(neg_inds) <= num_expected:
-            return neg_inds
+        if len(neg_inds) <= 0:
+            return False
```
Comment (ZwwWayne) [inline mmdet3d/core/bbox/samplers/iou_neg_piecewise_sampler.py:62]: This return is really weird here. It returns a different type. Should assert here

File: mmdet3d/core/bbox/samplers/iou_neg_piecewise_sampler.py [hunk 2]
```diff
@@ -87,12 +89,38 @@ def _sample_neg(self, assign_result, num_expected, **kwargs):
...
+                        if piece_neg_inds.numel() > 0:
+                            rand_idx = torch.randint(
+                                low=0,
+                                high=piece_neg_inds.numel(),
+                                size=(extend_neg_num, )).long()
+                            neg_inds_choice = torch.cat(
+                                [neg_inds_choice, piece_neg_inds[rand_idx]],
+                                dim=0)
+                        # if the numbers of nagetive samples == 0, we will
+                        # randomly select num_expected samples in all \
```
Comment (THU17cyz) [inline mmdet3d/core/bbox/samplers/iou_neg_piecewise_sampler.py:107]: remove \ ?

File: mmdet3d/core/bbox/samplers/iou_neg_piecewise_sampler.py [hunk 3]
```diff
@@ -58,8 +58,10 @@ def _sample_neg(self, assign_result, num_expected, **kwargs):
         neg_inds = torch.nonzero(assign_result.gt_inds == 0, as_tuple=False)
         if neg_inds.numel() != 0:
             neg_inds = neg_inds.squeeze(1)
-        if len(neg_inds) <= num_expected:
-            return neg_inds
+        if len(neg_inds) <= 0:
+            raise NotImplementedError(
+                'Not support to sample the negative samples when the length '
```
Comment (THU17cyz) [inline mmdet3d/core/bbox/samplers/iou_neg_piecewise_sampler.py:63]: Grammar error? Not support sampling?

**Review by THU17cyz — COMMENTED** [review]
Shall we add some unit tests?

**Comment by xiliu8006** [comment]
> Shall we add some unit tests?

the unittest is [here](https://github.com/open-mmlab/mmdetection3d/blob/9f0b01caf6aefed861ef4c3eb197c09362d26b32/tests/test_utils/test_samplers.py#L8)

**Review by THU17cyz — APPROVED** [review]

**Pull request merged**